Add ASOv1 to ASOv2 migration guide (#3898)
Azure/azure-service-operator: docs/hugo/content/guide/asov1-asov2-migration/_index.md

---
title: ASOv1 to ASOv2 migration guide
weight: 5
layout: single
cascade:
  - type: docs
  - render: always
---
ASOv1 is [deprecated](https://github.com/Azure/azure-service-operator/blob/main/README.md#aso-v1). 
We strongly recommend migrating to ASOv2 to continue getting new features and support.

## Assumptions

- Migrating a single cluster.
  - If you have ASOv1 installed on multiple clusters, apply these steps to each cluster.
- ASOv1 is installed in its default namespace `azureoperator-system`.

>   - If it's installed into a different namespace, replace all occurrences of `azureoperator-system` in this document with 

Reviewer: Suggest we avoid contractions as they are often considered informal.
```suggestion
  - If it is installed into a different namespace, replace all occurrences of `azureoperator-system` in this document with 
```

    the namespace it's installed into. 
- ASOv2 is not yet installed in the cluster.
- ASOv1 is managing a single subscription, so ASOv2 will also be managing that same single subscription via a single 
  global credential.
  - See [ASOv2 credential scope](../authentication/#credential-scope) for other credential scopes that allow for 
    managing multiple subscriptions and/or using multiple identities.

## Performing the migration

### Check the version of cert-manager

Ensure the version of cert-manager in your cluster supports `v1` resources.

```
kubectl api-resources --api-group cert-manager.io`
```
should show `v1` resources.

> ℹ️ **Note:** We strongly recommend ensuring that you're running the latest version of cert-manager 
> (1.14.4 at the time this article was written).

> ℹ️ **Note:** We also recommend ensuring that ASOv1 is configured to use `cert-manager.io/v1` resources.
> You can run `helm upgrade` and pass `--set certManagerResourcesAPIVersion=cert-manager.io/v1` to ensure ASOv1 is using

> > the v1 versions of the cert-manager resources.

Reviewer: Does this do an upgrade, or just return an error if non v1 versions are in use?

Author: It depends a bit on what you mean by "upgrade". It doesn't change the versions of any resources in the cluster (container images, helm charts, etc), at least assuming you use the same version of the ASOv1 chart as you already have installed. What it does is delete the ASOv1 `v1alpha2` cert-manager CRs and replace them with `v1` cert-manager CRs.

So the net effect of this is:
- Delete v1alpha2 cert manager custom resources (not the CRDs, the CRs)
- Create v1 cert manager custom resources w/ same configuration as above.

### Install ASOv2

Follow the [standard instructions](../../#installation). We recommend you use the same credentials as ASOv1 is currently using.

> ℹ️ **Note:** Make sure to follow the [guidelines](../crd-management/) for setting the `--crdPattern`. Configure ASOv2 
> to only manage the CRDs which you need.
> 

> > For example: if you're using storage accounts, resource groups, redis cache's, cosmosdbs, eventhubs, and SQL Azure, 

Reviewer: ```suggestion
> For example: if you are using storage accounts, resource groups, redis caches, cosmosdbs, eventhubs, and SQL Azure, 
```

> > then you'd configure 

Reviewer: ```suggestion
> then you would configure 
```

> `--set crdPattern='resources.azure.com/*;cache.azure.com/*;documentdb.azure.com/*;eventhub.azure.com/*;sql.azure.com/*;storage.azure.com/*'`

### Stop ASOv1 reconciliation

> Mark each ASOv1 resource with the `skipreconcile=true` annotation.

Reviewer 2: This must be hard to put annotation on _each_ resource. 
Should have a way in ASOv2 to pause reconciliation at the controller level for all resources, if not already there?

Author: I give commands just below of how to do it for a whole namespace at once. It's a one-liner so I don't think we need any other features. At most, we could make an `asoctl` feature to do the one-line like `asoctl annotate all` or something, but I don't think it's needed atm really.

> This annotation ensures that ASOv1 can no longer update or delete the resource in question in Azure.

Reviewer: ```suggestion
This annotation ensures that ASOv1 will no longer update or delete the resource in question in Azure.
```

To annotate a specific resource:
```

> kubectl annotate -n <ns> storageaccount <name> skipreconcile=true

Reviewer: Suggest being verbose about pieces that should be substituted by the reader

```suggestion
kubectl annotate -n <namespace> storageaccount <name> skipreconcile=true
```

Author: Done

```

To annotate all ASOv1 resources in a given namespace:
```

> kubectl annotate $(kubectl api-resources -o name | grep azure.microsoft.com | paste -sd "," - | xargs kubectl get -A -o name) -n ns1 skipreconcile=true`

Reviewer: I think you missed a `<ns>` substitution mark

```suggestion
kubectl annotate $(kubectl api-resources -o name | grep azure.microsoft.com | paste -sd "," - | xargs kubectl get -A -o name) -n <namespace> skipreconcile=true`
```

Also, does the inner `kubectl api-resources` require a namespace filter?

Author: It does not. Fixed the ns1 bit.

```

> Once you've annotated the resources, double-check that they all have the `skipreconcile` annotation. 

Reviewer: ```suggestion
Once you have annotated the resources, double-check that they all have the expected `skipreconcile` annotation. 
```

> > ⚠️ **Important**: If you delete an ASOv1 resource that does not have this annotation it will delete the underlying resource 
> > in Azure and may cause downtime or an outage

Reviewer: ```suggestion
> ⚠️ **Important**: If you delete an ASOv1 resource that does not have this annotation it will delete the underlying Azure resource, 
> which may cause downtime or an outage.
```

> ### Use `asoctl` to import the resources from Azure into ASOv2
> 
> Download the latest version of [asoctl](../../tools/asoctl/).

Reviewer: Need to define what `asoctl` is.

```suggestion
### Use `asoctl` to import the resources from Azure into ASOv2

The commandline tool `asoctl` is a utility for working with ASO. Here we are going to use it to convert your existing Azure resources for use with ASOv2. 

Download the latest version of [asoctl](../../tools/asoctl/).
```

Author: I slightly rephrased your suggestion to:
> The commandline tool `asoctl` is a utility for working with ASO. Here we are going to use it to import your existing 
Azure resources into ASOv2.

Export the resources you want to transfer using the `asoctl import azure-resource` command. `import azure-resource`
imports the referenced resource and all its child resources by default. This means if you want to import the resources 
from a specific resource group being managed by ASOv1 you can just refer to the resource group.

> When running this command it's recommended you use at least asoctl v2.7.0 and pass the 
> `-n` (namespace) and `-a` (annotation) arguments to import the resources to a particular namespace with 
> the `serviceoperator.azure.com/reconcile-policy: skip` annotation.

Reviewer: In documentation, I prefer to use the long form of flags for clarity - users can substitute the short form if they choose.

```suggestion
When running this command it's recommended you use at least asoctl v2.7.0 and pass the 
`--namespace` (or `-n`) and `--annotation` (or `-a`) arguments to import the resources to a particular namespace with 
the `serviceoperator.azure.com/reconcile-policy: skip` annotation.
```

Author: Fair enough, fixed

```

> asoctl import azure-resource /subscriptions/<subid>/resourceGroups/<rg> -o resources.yaml -n ns1 -a serviceoperator.azure.com/reconcile-policy=skip

Reviewer: ```suggestion
asoctl import azure-resource /subscriptions/<subid>/resourceGroups/<rg> -o resources.yaml --namespace ns1 --annotation serviceoperator.azure.com/reconcile-policy=skip
```

Reviewer: Also, may be worth pointing out the `--output-folder` option, which writes each resource to a separate file. Useful if you have more than a trivial number of resources.

Author: Fixed the args and added a blurb about `--output-folder`

```

> This should produce a file that looks something like this (this is a simplified file showing just a resource group and

Reviewer: Suggested rephrase to avoid sequential _this_

```suggestion
This should produce a file similar to the one shown here (this is a simplified file showing just a resource group and
```

a single resource for the sake of brevity. It's likely that your file is much larger):

```yaml
---
apiVersion: resources.azure.com/v1api20200601
kind: ResourceGroup
metadata:
  name: aso-test
  namespace: ns1
  annotations:
    serviceoperator.azure.com/reconcile-policy: skip
spec:
  azureName: aso-test
  location: westus
---
apiVersion: storage.azure.com/v1api20230101
kind: StorageAccount
metadata:
  name: aso-test-storage1
  namespace: ns1
  annotations:
    serviceoperator.azure.com/reconcile-policy: skip
spec:
  accessTier: Hot
  allowBlobPublicAccess: false
  allowCrossTenantReplication: false
  azureName: asoteststorage1
  encryption:
    keySource: Microsoft.Storage
    services:
      blob:
        enabled: true
        keyType: Account
      file:
        enabled: true
        keyType: Account
  kind: StorageV2
  location: westus
  minimumTlsVersion: TLS1_0
  networkAcls:
    bypass: AzureServices
    defaultAction: Allow
  owner:
    name: aso-test
  sku:
    name: Standard_RAGRS
    tier: Standard
  supportsHttpsTrafficOnly: true
```

### Examine `resources.yaml` to ensure it has the resources you expect

> Once you've got `resources.yaml` locally, examine it to ensure that it has the resources you expect. Depending on the size of

Reviewer: ```suggestion
Once you have `resources.yaml` locally, examine it to ensure that it has the resources you expect. Depending on the size of
```

ASOv1 generated these passwords for you, but in ASOv2 you control these passwords and their lifecycle (including rollover).

`asoctl` does not automatically include secret password fields in its output. Instead, you must supply a Kubernetes
secret containing the password which ASOv2 will supply to Azure.

The following ASOv1 resources automatically generated usernames and passwords which you will need to take ownership of
when migrating to ASOv2:
- Azure SQL Server
- Azure SQL User
- My SQL Server
- My SQL User
- PostgreSQL SQL Server
- PostgreSQL SQL User
- VM
- VMSS

The process for each of these resources is the same:
1. Find the ASOv1 managed secret which contains the current password.
2. Create a new secret containing the same password.
3. Update the ASOv2 `resources.yaml` to supply the new password in the correct location.

See [examples](#examples) for examples of how to do this with various resource types.

### Configure `resources.yaml` to have annotation [serviceoperator.azure.com/reconcile-policy: skip](../annotations/#serviceoperatorazurecomreconcile-policy)

This can be done automatically by `asoctl` during resource import with the 

> `-a serviceoperator.azure.com/reconcile-policy=skip` argument or manually in the `resources.yaml` afterward.

Reviewer: ```suggestion
`--annotation serviceoperator.azure.com/reconcile-policy=skip` argument or manually by modifying `resources.yaml` afterward.
```

> ⚠️ **Important:** Be _very_ careful issuing deletions of ASOv2 resources once you've imported them if they do not have 
> the serviceoperator.azure.com/reconcile-policy=skip annotation. Just like ASOv1,
> ASOv2 will delete the underlying Azure resource by default!

> > It's _strongly_ recommended that you use asoctl's `-a serviceoperator.azure.com/reconcile-policy=skip` flag and only remove
> > that annotation a few resources at a time to ensure things are working.

Reviewer: ```suggestion
> It is _strongly_ recommended that you use asoctl's `--annotation serviceoperator.azure.com/reconcile-policy=skip` flag and only remove
> that annotation a few resources at a time to ensure things are working.
```
